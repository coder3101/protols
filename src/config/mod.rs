use serde::{Deserialize, Serialize};

pub mod workspace;

fn default_clang_format_path() -> String {
    "clang-format".to_string()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct ProtolsConfig {
    pub config: Config,
    pub formatter: FormatterConfig,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FormatterConfig {
    #[serde(default = "default_clang_format_path")]
    pub clang_format_path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct Config {
    pub include_paths: Vec<String>,
    pub single_file_mode: bool,
    pub disable_parse_diagnostics: bool,
    pub experimental: ExperimentalConfig,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct ExperimentalConfig {
    pub use_protoc_diagnostics: bool,
}

impl Default for FormatterConfig {
    fn default() -> Self {
        Self {
            clang_format_path: default_clang_format_path(),
        }
    }
}
