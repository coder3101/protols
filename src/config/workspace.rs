use std::{
    collections::{HashMap, HashSet},
    env,
    path::{Path, PathBuf},
};

use async_lsp::lsp_types::{Url, WorkspaceFolder};
use pkg_config::Config;

use crate::formatter::clang::ClangFormatter;

use super::ProtolsConfig;

const CONFIG_FILE_NAMES: [&str; 2] = [".protols.toml", "protols.toml"];

pub struct WorkspaceProtoConfigs {
    workspaces: HashSet<Url>,
    configs: HashMap<Url, ProtolsConfig>,
    formatters: HashMap<Url, ClangFormatter>,
    protoc_include_prefix: Vec<PathBuf>,
    cli_include_paths: Vec<PathBuf>,
    init_include_paths: Vec<PathBuf>,
}

impl WorkspaceProtoConfigs {
    pub fn new(cli_include_paths: Vec<PathBuf>) -> Self {
        // Try to find protobuf library and get its include paths
        // Do not emit metadata on stdout as LSP programs can consider
        // it part of spec
        let protoc_include_prefix = Config::new()
            .atleast_version("3.0.0")
            .env_metadata(false)
            .cargo_metadata(false)
            .probe("protobuf")
            .map(|lib| lib.include_paths)
            .unwrap_or_default();

        Self {
            workspaces: HashSet::new(),
            formatters: HashMap::new(),
            configs: HashMap::new(),
            protoc_include_prefix,
            cli_include_paths,
            init_include_paths: Vec::new(),
        }
    }

    fn get_config_file_path(wpath: &PathBuf) -> Option<PathBuf> {
        for file in CONFIG_FILE_NAMES {
            let p = Path::new(&wpath).join(file);
            match std::fs::exists(&p) {
                Ok(exists) if exists => return Some(p),
                _ => continue,
            }
        }
        None
    }

    pub fn add_workspace(&mut self, w: &WorkspaceFolder) {
        let Ok(wpath) = w.uri.to_file_path() else {
            return;
        };

        let path = Self::get_config_file_path(&wpath).unwrap_or_default();
        let content = std::fs::read_to_string(path).unwrap_or_default();

        let wr: ProtolsConfig = basic_toml::from_str(&content).unwrap_or_default();
        let fmt = ClangFormatter::new(
            &wr.config.path.clang_format,
            Some(wpath.to_str().expect("non-utf8 path")),
        );

        self.workspaces.insert(w.uri.clone());
        self.configs.insert(w.uri.clone(), wr);
        self.formatters.insert(w.uri.clone(), fmt);
    }

    pub fn get_config_for_uri(&self, u: &Url) -> Option<&ProtolsConfig> {
        self.get_workspace_for_uri(u)
            .and_then(|w| self.configs.get(w))
    }

    pub fn get_formatter_for_uri(&self, u: &Url) -> Option<&ClangFormatter> {
        self.get_workspace_for_uri(u)
            .and_then(|w| self.formatters.get(w))
    }

    pub fn get_workspace_for_uri(&self, u: &Url) -> Option<&Url> {
        let upath = u.to_file_path().ok()?;
        self.workspaces
            .iter()
            .find(|&k| upath.starts_with(k.to_file_path().unwrap()))
    }

    pub fn set_init_include_paths(&mut self, paths: Vec<PathBuf>) {
        self.init_include_paths = paths;
    }

    pub fn get_include_paths(&self, uri: &Url) -> Option<Vec<PathBuf>> {
        let cfg = self.get_config_for_uri(uri)?;
        let w = self.get_workspace_for_uri(uri)?.to_file_path().ok()?;

        let mut ipath: Vec<PathBuf> = cfg
            .config
            .include_paths
            .iter()
            .map(PathBuf::from)
            .map(|p| if p.is_relative() { w.join(p) } else { p })
            .collect();

        // Add CLI include paths
        for path in &self.cli_include_paths {
            if path.is_relative() {
                ipath.push(w.join(path));
            } else {
                ipath.push(path.clone());
            }
        }

        // Add initialization include paths
        for path in &self.init_include_paths {
            if path.is_relative() {
                ipath.push(w.join(path));
            } else {
                ipath.push(path.clone());
            }
        }

        ipath.push(w.to_path_buf());
        ipath.extend_from_slice(&self.protoc_include_prefix);
        Some(ipath)
    }

    pub fn no_workspace_mode(&mut self) {
        let wr = ProtolsConfig::default();
        let rp = if cfg!(target_os = "windows") {
            let mut d = String::from("C");
            if let Ok(cdir) = env::current_dir()
                && let Some(drive) = cdir.components().next()
            {
                d = drive.as_os_str().to_string_lossy().to_string()
            }
            format!("{d}://")
        } else {
            String::from("/")
        };
        let uri = match Url::from_file_path(&rp) {
            Err(err) => {
                tracing::error!(?err, "failed to convert path: {rp} to Url");
                return;
            }
            Ok(uri) => uri,
        };

        let fmt = ClangFormatter::new(&wr.config.path.clang_format, None);

        self.workspaces.insert(uri.clone());
        self.configs.insert(uri.clone(), wr);
        self.formatters.insert(uri.clone(), fmt);
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{Url, WorkspaceFolder};
    use insta::assert_yaml_snapshot;
    use std::path::PathBuf;
    use tempfile::tempdir;

    use super::{CONFIG_FILE_NAMES, WorkspaceProtoConfigs};

    #[test]
    fn test_get_for_workspace() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let tmpdir2 = tempdir().expect("failed to create temp2 directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        let mut ws = WorkspaceProtoConfigs::new(vec![]);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir2.path()).unwrap(),
            name: "Test2".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let outworkspace =
            Url::from_file_path(tempdir().unwrap().path().join("out.proto")).unwrap();
        let inworkspace2 = Url::from_file_path(tmpdir2.path().join("foobar.proto")).unwrap();

        assert!(ws.get_config_for_uri(&inworkspace).is_some());
        assert!(ws.get_config_for_uri(&inworkspace2).is_some());
        assert!(ws.get_config_for_uri(&outworkspace).is_none());

        assert!(ws.get_workspace_for_uri(&inworkspace).is_some());
        assert!(ws.get_workspace_for_uri(&inworkspace2).is_some());
        assert!(ws.get_workspace_for_uri(&outworkspace).is_none());

        assert_yaml_snapshot!(ws.get_config_for_uri(&inworkspace).unwrap());
        assert_yaml_snapshot!(ws.get_config_for_uri(&inworkspace2).unwrap());
    }

    #[test]
    fn test_get_formatter_for_uri() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let tmpdir2 = tempdir().expect("failed to create temp2 directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        let mut ws = WorkspaceProtoConfigs::new(vec![]);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir2.path()).unwrap(),
            name: "Test2".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let outworkspace =
            Url::from_file_path(tempdir().unwrap().path().join("out.proto")).unwrap();
        let inworkspace2 = Url::from_file_path(tmpdir2.path().join("foobar.proto")).unwrap();

        assert!(ws.get_formatter_for_uri(&outworkspace).is_none());
        assert_eq!(
            ws.get_formatter_for_uri(&inworkspace).unwrap().path,
            "/usr/bin/clang-format"
        );
        assert_eq!(
            ws.get_formatter_for_uri(&inworkspace2).unwrap().path,
            "clang-format"
        );
    }

    #[test]
    fn test_loading_different_config_files() {
        let tmpdir = tempdir().expect("failed to create temp directory");

        for file in CONFIG_FILE_NAMES {
            let f = tmpdir.path().join(file);
            std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

            let mut ws = WorkspaceProtoConfigs::new(vec![]);
            ws.add_workspace(&WorkspaceFolder {
                uri: Url::from_directory_path(tmpdir.path()).unwrap(),
                name: "Test".to_string(),
            });

            // check we really loaded the config file
            let workspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
            assert!(ws.get_workspace_for_uri(&workspace).is_some());
        }
    }

    #[test]
    fn test_cli_include_paths() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        // Set CLI include paths
        let cli_paths = vec![
            PathBuf::from("/path/to/protos"),
            PathBuf::from("relative/path"),
        ];
        let mut ws = WorkspaceProtoConfigs::new(cli_paths);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let include_paths = ws.get_include_paths(&inworkspace).unwrap();

        // Check that CLI paths are included in the result
        assert!(
            include_paths
                .iter()
                .any(|p| p.ends_with("relative/path") || p == &PathBuf::from("/path/to/protos"))
        );

        // The relative path should be resolved relative to the workspace
        let resolved_relative_path = tmpdir.path().join("relative/path");
        assert!(include_paths.contains(&resolved_relative_path));

        // The absolute path should be included as is
        assert!(include_paths.contains(&PathBuf::from("/path/to/protos")));
    }

    #[test]
    fn test_init_include_paths() {
        let tmpdir = tempdir().expect("failed to create temp directory");
        let f = tmpdir.path().join("protols.toml");
        std::fs::write(f, include_str!("input/protols-valid.toml")).unwrap();

        // Set both CLI and initialization include paths
        let cli_paths = vec![PathBuf::from("/cli/path")];
        let init_paths = vec![
            PathBuf::from("/init/path1"),
            PathBuf::from("relative/init/path"),
        ];

        let mut ws = WorkspaceProtoConfigs::new(cli_paths);
        ws.set_init_include_paths(init_paths);
        ws.add_workspace(&WorkspaceFolder {
            uri: Url::from_directory_path(tmpdir.path()).unwrap(),
            name: "Test".to_string(),
        });

        let inworkspace = Url::from_file_path(tmpdir.path().join("foobar.proto")).unwrap();
        let include_paths = ws.get_include_paths(&inworkspace).unwrap();

        // Check that initialization paths are included
        assert!(include_paths.contains(&PathBuf::from("/init/path1")));

        // The relative path should be resolved relative to the workspace
        let resolved_relative_path = tmpdir.path().join("relative/init/path");
        assert!(include_paths.contains(&resolved_relative_path));

        // CLI paths should still be included
        assert!(include_paths.contains(&PathBuf::from("/cli/path")));
    }
}
