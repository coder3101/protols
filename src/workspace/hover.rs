use std::path::{Path, PathBuf};

use async_lsp::lsp_types::{MarkupContent, MarkupKind};

use crate::{
    context::hoverable::Hoverables, docs, state::ProtoLanguageState,
    utils::split_identifier_package,
};

fn format_import_path_hover_text(path: &str, p: &Path) -> String {
    format!(
        r#"Import: `{path}` protobuf file,
---
Included from {}"#,
        p.to_string_lossy()
    )
}

fn format_identifier_hover_text(identifier: &str, package: &str, result: &str) -> String {
    format!(
        r#"`{identifier}` message or enum type, package: `{package}`
---
{result}"#
    )
}

impl ProtoLanguageState {
    pub fn hover(
        &self,
        ipath: &[PathBuf],
        curr_package: &str,
        hv: Hoverables,
    ) -> Option<MarkupContent> {
        let v = match hv {
            Hoverables::FieldType(field) => docs::BUITIN
                .get(field.as_str())
                .map(ToString::to_string)
                .unwrap_or_default(),

            Hoverables::ImportPath(path) => ipath
                .iter()
                .map(|p| p.join(&path))
                .find(|p| p.exists())
                .map(|p| format_import_path_hover_text(&path, &p))
                .unwrap_or_default(),

            Hoverables::Identifier(identifier) => {
                let (mut package, identifier) = split_identifier_package(identifier.as_str());
                if package.is_empty() {
                    package = curr_package;
                }

                // Identifier is user defined type or well known type

                // If well known types, check in wellknown docs,
                // otherwise check in trees
                match docs::WELLKNOWN
                    .get(format!("{package}.{identifier}").as_str())
                    .map(|&s| s.to_string())
                {
                    Some(res) => res,
                    None => {
                        let mut trees = vec![];

                        // If package != curr_package, either identifier is from a completely new package
                        // or relative package from within. As per name resolution first resolve relative
                        // packages, add all relative trees in search list
                        if curr_package != package {
                            let fullpackage = format!("{curr_package}.{package}");
                            trees.append(&mut self.get_trees_for_package(&fullpackage));
                        }

                        // Add all direct package trees
                        trees.append(&mut self.get_trees_for_package(package));

                        // Find the first field hovered in the trees
                        let res = trees.iter().find_map(|tree| {
                            let content = self.get_content(&tree.uri);
                            let res = tree.hover(identifier, content);
                            if res.is_empty() {
                                None
                            } else {
                                Some(res[0].clone())
                            }
                        });

                        // Format the hover text and return
                        // TODO: package here is literally what was hovered, incase of
                        // relative it is only the relative part, should be full path, should
                        // probably figure out the package from the tree which provides hover and
                        // pass here.
                        res.map(|r| format_identifier_hover_text(identifier, package, &r))
                            .unwrap_or_default()
                    }
                }
            }
        };

        match v {
            v if v.is_empty() => None,
            v => Some(MarkupContent {
                kind: MarkupKind::Markdown,
                value: v,
            }),
        }
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use crate::config::Config;
    use crate::context::hoverable::Hoverables;
    use crate::state::ProtoLanguageState;
    #[test]
    fn workspace_test_hover() {
        let ipath = vec![std::env::current_dir().unwrap().join("src/workspace/input")];
        let a_uri = "file://input/a.proto".parse().unwrap();
        let b_uri = "file://input/b.proto".parse().unwrap();
        let c_uri = "file://input/c.proto".parse().unwrap();

        let a = include_str!("input/a.proto");
        let b = include_str!("input/b.proto");
        let c = include_str!("input/c.proto");

        let mut state: ProtoLanguageState = ProtoLanguageState::new();
        state.upsert_file(&a_uri, a.to_owned(), &ipath, 3, &Config::default(), false);
        state.upsert_file(&b_uri, b.to_owned(), &ipath, 2, &Config::default(), false);
        state.upsert_file(&c_uri, c.to_owned(), &ipath, 2, &Config::default(), false);

        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::Identifier("google.protobuf.Any".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::Identifier("Author".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::FieldType("int64".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::Identifier("Author.Address".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::Identifier("com.utility.Foobar.Baz".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.utility",
            Hoverables::Identifier("Baz".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.workspace",
            Hoverables::Identifier("com.inner.Why".to_string())
        ));
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.inner",
            Hoverables::Identifier(".com.inner.secret.SomeSecret".to_string())
        ));
        // relative path hover
        assert_yaml_snapshot!(state.hover(
            &ipath,
            "com.inner",
            Hoverables::Identifier("secret.SomeSecret".to_string())
        ))
    }
}
