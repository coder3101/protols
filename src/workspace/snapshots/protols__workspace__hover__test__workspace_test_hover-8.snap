---
source: src/workspace/hover.rs
expression: "state.hover(&ipath, \"com.inner\",\nHoverables::Identifier(\".com.inner.secret.SomeSecret\".to_string()))"
---
kind: markdown
value: "`SomeSecret` message or enum type, package: `com.inner.secret`\n---\nSomeSecret is a real secret with hidden string"
